### PR TITLE
fix(block): Fixed title_style not rendered

### LINKED ISSUE #349: Block title_style is not rendered
## Description
In the [recent PR to support multiple titles](https://github.com/ratatui-org/ratatui/pull/232), we accidentally made title_style no longer effective. It's useful to have this as a fallback style as it makes it easy to apply a style to all titles except one:
e.g. 
```rust
Block::new()
  .title_style(Style::new().bold())
  .title(Title::from("foo").alignment(Alignment::Left))
  .title(Title::from("bar").alignment(Alignment::Center))
  .title(Title::from("foo".not_bold().dim()).alignment(Alignment::Right))
```

## To Reproduce
```rust
    frame.render_widget(
        Block::new()
            .borders(Borders::all())
            .title("Styled Title")
            .title_style(Style::new().yellow()),
        area1,
    );
    frame.render_widget(
        Block::new()
            .borders(Borders::all())
            .title(Title::from(Line::styled(
                "Styled Line",
                Style::new().yellow(),
            ))),
        area 2,
    );
```

## Expected behavior
The Block's title style should be applied to the title.
The order which styles should effect the title (later styles should override earlier):
1. Any style applied to the surrounding area
2. `Block.style`
3. `Block.border_style`
4. `Block.title_style` - this is the missing piece
5. `Title.content.style` (`Line.style`): Out of scope for this issue as `Line` currently doesn't have a style, but should
6. `Title.content.spans` -> `Span.style`

## Screenshots
image

## Environment
- OS: Mac
- Terminal Emulator: iterm2
- Font: FiraCode NF
- Crate version: 0.22
- Backend: crossterm

## Additional context

This is the code that needs to be modified for this:

https://github.com/ratatui-org/ratatui/blob/bf4944683d6afb6f42bec80a1bd308ecdac50cbc/src/widgets/block.rs#L418-L423
https://github.com/ratatui-org/ratatui/blob/bf4944683d6afb6f42bec80a1bd308ecdac50cbc/src/widgets/block.rs#L444-L449
https://github.com/ratatui-org/ratatui/blob/bf4944683d6afb6f42bec80a1bd308ecdac50cbc/src/widgets/block.rs#L465-L470

There's a simple ways to approach this (just create a new line from the old line with the appropriate title_style applied before the line's style).

There's also a more complex fix for this that changes the way line style is stored / rendered so that line style overrides can be consistently handled by any widget. If you decide to implement the complex approach, please implement the simple approach first so there is tests for the intended behavior, and discuss the API changes for this prior to implementation as this likely requires changes to the public api of ratatui.

### PR DESCRIPTION
**Pull request by Valentin271**
This PR fixes #349 the easy way (see issue). I think the _complex fix_ requires another separate PR.

I added a unit test and made sure it didn't pass prior to this modification. I added a lot of test cases because there are three separate places where this bug can occur.

I didn't change the corresponding example to demonstrate the behavior as there is already an [opened PR](https://github.com/ratatui-org/ratatui/pull/351) that will likely be merged.

I didn't want to create a new spans (or lines) because in the future they might contain more than just `content` and `style` and this would lead to some half cloned spans. Thus I preferred cloning and modifying them locally.

### CODE REVIEW

File: src/widgets/block.rs [hunk 1]
```diff
@@ -415,10 +415,16 @@ impl<'a> Block<'a> {
                 let title_x = current_offset;
                 current_offset += title.content.width() as u16 + 1;
 
+                // Clone the title's content, applying block title style then the title style
+                let mut content = title.content.clone();
+                for sp in content.spans.iter_mut() {
+                    sp.style = self.titles_style.patch(sp.style);
```
Comment (joshka) [inline src/widgets/block.rs:421]: Small nitpick: I'd generally avoid abbreviations when possible.

File: src/widgets/block.rs [hunk 2]
```diff
@@ -917,4 +936,50 @@ mod tests {
             assert_buffer_eq!(buffer, Buffer::with_lines(vec![expected]));
         }
     }
+
+    #[test]
+    fn title_style_overrides_block_title_style() {
```
Comment (joshka) [inline src/widgets/block.rs:941]: I have a strong preference for keeping tests super simple. I found myself scrolling up to the test data and back down to the test itself to cross reference the data and behavior. We're testing 4 things that matter over 45 lines, which can be expressed as perhaps 4 tests each ~9 lines. The main benefit of splitting things this way is that the tests can help document the code. E.g. what about something like:

```rust
    #[test]
    fn title_content_style_renders_style() {
        for alignment in [Alignment::Left, Alignment::Center, Alignment::Right] {
            let mut buffer = Buffer::empty(Rect::new(0, 0, 4, 1));
            Block::default()
                .title_alignment(alignment)
                .title("test".yellow())
                .render(buffer.area, &mut buffer);
            let mut expected = Buffer::with_lines(vec!["test"]);
            expected.set_style(expected.area, Style::new().yellow());
            assert_buffer_eq!(buffer, expected);
        }
    }
```
etc.

**Review by joshka — APPROVED** [review]
Thanks for the PR and welcome to Ratatui.

Generally LGTM. I had a nitpick and a preference that has to do with keeping tests simple. Can be merged as-is or updated to address the feedback (both are not huge points).

File: src/widgets/block.rs (hunk 1 above)
Comment (Valentin271) [reply src/widgets/block.rs]: Fixed :)

File: src/widgets/block.rs (hunk 2 above)
Comment (Valentin271) [reply src/widgets/block.rs]: I was was testing basically 3 different cases, there are now separated in 3 tests and are easier to understand.

**Review by joshka — APPROVED** [review]
LGTM - Thanks for the fix. We appreciate your effort.

**Comment by joshka** [comment]
Merging this even though checks are failing as the failures are due to the MSRV issue.

**Pull request merged**